### PR TITLE
Added Verification for GPU Configuration using Native Tensor Cores (mma.sync)

### PR DESCRIPTION
**Pull request by manishucsd**
This PR add verification for GPU configurations using native Tensor Cores. 

Additionally, improves the existing code verifying SIMT and Tensor Core (wmma) pipelines. There was duplication of checks for SIMT and Tensor Cores. This PR structures the moving parts such as difference instructions shapes for SIMT, WMMA, MMA.SYNC into a function and moves the common checks at one place. The additional alignment constraints placed upon WMMA and MMA.SYNC are checked only for the Tensor Core pipelines. Finally, the PR makes an efforts to improve the readability of constraints.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 1]
```diff
@@ -4,225 +4,265 @@
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+#include <iostream>
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: remove

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 2]
```diff
@@ -4,225 +4,265 @@
...
+                                 rhsType.cast<ShapedType>().getElementType(),
+                                 threadBlockShape, softwarePipelineDepth);
+
+
+  if (sharedMemorySize > kNvidiaSM80SharedMemCapacity) {
+    return op->emitError("Shared memory size ")
+           << sharedMemorySize << " exceeds the limit of "
+           << kNvidiaSM80SharedMemCapacity;
   }
+#endif
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: please remove the code within `#if0` you can leave the TODO only.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 3]
```diff
@@ -4,225 +4,265 @@
...
-
 namespace mlir {
 namespace iree_compiler {
 
+////////////////////////////////////////////////////////////////////////////////
+/// Constants used in the matmul lowering verifiers.
 constexpr unsigned kWorkgroupTileLevel = 0;
-constexpr int kSharedMemSizeBytes = 64 * 1024;
+constexpr int kNvidiaSM86SharedMemCapacity = 128 * 1024;
+constexpr int kNvidiaSM80SharedMemCapacity = 192 * 1024;
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: this is unused, let's remove it.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 4]
```diff
@@ -4,225 +4,265 @@
...
+}
 
-  return success();
+/// Returns the Shared Memory Size in bytes.
+static int getSharedMemorySizeInBytes(Type aElementType, Type bElementType,
+                                      SmallVector<int64_t> threadBlockShape,
+                                      int64_t pipelineDepth) {
+  // TODO: Implement this function (@KoolJBlack is working on it).
+  return 0;
 }
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: remove

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 5]
```diff
@@ -4,225 +4,265 @@
...
-               << pipelineName;
+    case IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUMatmulTensorCore:
+      // Tensor Core Pipeline / WMMA API
+      // See WMMA Shapes here
+      if (inputElementType.isF16() || inputElementType.isBF16()) {
+        instructionShape = {16, 16, 16};
+      } else if (inputElementType.isF32()) {
+        instructionShape = {16, 16, 8};
+      } else {
+        llvm_unreachable("expected f16, bf16 or f32 for tensor core pipeline");
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: make those  assert

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 6]
```diff
@@ -4,225 +4,265 @@
...
-    IREE::Codegen::TranslationInfoAttr translationInfo,
-    ArrayRef<int64_t> workgroupSize) {
-  auto pipeline =
-      IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUMatmulSimt;
-  StringRef pipelineName = stringifyEnum(pipeline);
-  if (workgroupSize.empty()) {
-    return op->emitOpError("expected workgroup size for GPU pipelines");
-  }
+/// Use the constexpr to convey the meaning of the indices.
+// Dimenstions identifiers for: worgroup size(x, y, z), and thread(x, y, z)
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: nit: use consistent number of `/` for comments.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 7]
```diff
@@ -4,225 +4,265 @@
...
 
-  unsigned softwarePipelinedepth = translationInfo.getSoftwarePipelineDepth();
-  StringRef pipelineName = stringifyEnum(pipeline);
-  if (workgroupSize.empty()) {
-    return op->emitOpError("expected workgroup size for GPU pipelines");
-  }
+  // Get compilation pipeline.
+  auto pipeline = translationInfo.getPassPipeline();
+  StringRef pipelineName = stringifyEnum(pipeline.getValue());
+  std::cout << "pipelineName: " << pipelineName.str() << std::endl;
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: remove

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 8]
```diff
@@ -4,225 +4,265 @@
...
-               "Software pipelining is not supported when first level K tile "
-               "size is same as matrix reduction size.\n This dispatch has\nk "
-               "tile: ")
-           << firstLevelTileSizes[2]
-           << "\nMatrix reduction size: " << lhsShape[1]
-           << "\nPipelinedepth: " << softwarePipelinedepth;
+  // Number of warps in matmul problem dimension M, N, and K.
+  // Note the following thread dim -> problem dim mapping:
+  // DimY -> ProblemDimM, DimX -> ProblemDimN, DimZ -> ProblemDimK.
+  SmallVector<int64_t> numWarps{workgroupSize[kDimY],
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: having the num warps be in different order than workgroup size is a confusing. Number of warps should be using X, Y, Z not M, N, K order

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 9]
```diff
@@ -4,225 +4,265 @@
...
+  std::cout << "matmulShape: " << matmulShape[0] << "x" << matmulShape[1] << "x"
+            << matmulShape[2] << std::endl;
+  std::cout << "threadBlockShape: " << threadBlockShape[0] << "x"
+            << threadBlockShape[1] << "x" << threadBlockShape[2] << std::endl;
+  std::cout << "warpShape: " << warpShape[0] << "x" << warpShape[1] << "x"
+            << warpShape[2] << std::endl;
+  std::cout << "instructionShape: " << instructionShape[0] << "x"
+            << instructionShape[1] << "x" << instructionShape[2] << std::endl;
+#endif
+
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: remove, please make sure to clean up PRs before sending them for review.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 10]
```diff
@@ -4,225 +4,265 @@
...
-      secondLevelTileSizes[1] % tensorCoreSize[1] != 0 ||
-      secondLevelTileSizes[2] % tensorCoreSize[2] != 0) {
-    return op->emitOpError(
-               "tensorcore size doesn't factor into second level tile size "
-               "for ")
-           << pipelineName;
+  // Verify the number of warps in the problem K dimension is 1.
+  if (numWarps[kK] != 1) {
+    return op->emitError("Number of warps in problem K dimension ")
+           << numWarps[kK] << " is not 1";
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: this doesn't make sense to me. There is no such thing as numWarps along K.

**Comment by ThomasRaoux** [comment]
Please add a test

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp (hunk 8 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: The figure below is the convention and mapping I followed:

Screenshot 2023-03-22 at 5 48 05 PM

At some level we need to map arch dims (x,y,z) to the problem. I applied and followed problem (M,N,K) mapping for everything below the dotted line:

`grid = {bx, by, bz}`
`block/workGroup = {tx, ty, tz}`
--------------------- map architecture's (x,y,z) dims to matmul problem (m, n, k) ------------------------------
`numWarp = {num of warps in [kM], num of warps  in [kN], num of warps in  in [kK]} =  {ty, tx/32, tz}`

`threadBlockShape = {num of tiles in [kM], num of tiles in [kN], num of tiles in [kK]}`

Having the dotted line above numWarp a.k.a. applying the mapping from x,y,z arch dim to m,n,k problem allows nice use of kM for both numerator and denominator in the formula below. The mapping part is happening in the computation of `numWarp`
`warpShape = {threadBlockShape[kM]/numWarp[kM], 
                           threadBlockShape[kN]/numWarp[kN],
                           threadBlockShape[kK]/numWarp[kK]}` 
                           
`instructionShape = {num elements in M, num elements in N, num elements in N}` // using m,n,k here is straight forward as it maps nicely to the use of m,n,k in mma.sync*m16n8k16.


Also, I want to mention that to completely avoid this confusion with threads in x and y dim, one option would be to launch threads in just x-dim and create a layout of threads. It maybe out of scope for now, but to support all different layouts say row-row, row-col, col-row, and row-col just having threads in one dim and creating explicit layout gives more control. It also allows having different thread layout for different part of the problem (mainloop vs. epilogue) in the same kernel. 

It is just a choice and whatever is easier visualize, we can move the dotted line below numWarp and apply mapping when computing `warpShape`.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp (hunk 10 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: number of warps in the K dim is 1. If we have more than one warp to accumulate threadBlockShape[kK] dim then what we have is `Sliced-K` kernel equivalent from CUTLASS. 

`Split-K` is dividing the ProblemK across multiple thread blocks, and `Sliced-K` is dividing threadBlockShape[kK]  across multiple warps. We don't have `Sliced-K` and we don't plan to support it. 

To visualize this in the picture above, the green lines shows the warp shape (2x2x1 in the picture), The green lines are only in matrixC (threadBlocShape[kM] x threadBlocShape[kN)]  and not in matrixA or matrixB threadBlockShape[kK]. 

This condition is to check, if the user of `OpenXLA/iree` do not end up providing a compilation config where numWarp > 1.

Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: This condition is to check, if the user has given tile config such that it ends up having numWarp > 1, we catch that.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp (hunk 8 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: > It is just a choice and whatever is easier visualize, we can move the dotted line below numWarp and apply mapping when computing warpShape.

That would feel more natural to me as having a number of warp in M, N, K doesn't make sense to me. The number of warp is analogue to number of thread. warpShape feels like the right place to handle the logic.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 11]
```diff
@@ -12,217 +12,231 @@
...
+           << workgroupSize[kDimZ] << " with compilation pipeline "
+           << pipelineName;
   }
 
-  // Verify shared memory usage of operands after tiling requires <= 64Kb
-  // combined space.
-  unsigned bytesSize =
-      inputType.cast<ShapedType>().getElementType().getIntOrFloatBitWidth() / 8;
+  // Verify shared memory usage is within the limit.
+  // TODO: (@KoolJBlack) is working on adding this check.
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp]: nit: the format followed is usually:
`TODO(KoolJBlack): ...`

**Review by ThomasRaoux — APPROVED** [review]
Looks good to me but I would change the numWarp mapping as it looks confusing in my opinion.

**Review by ThomasRaoux — APPROVED** [review]

**Pull request merged**